// @aws-sdk/client-codeartifact/dist-cjs/models/models_0.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PublishPackageVersionRequestFilterSensitiveLog = exports.GetPackageVersionAssetResultFilterSensitiveLog = exports.PackageVersionSortType = exports.PackageVersionOriginType = exports.DomainStatus = exports.PackageVersionStatus = exports.PackageVersionErrorCode = exports.ValidationException = exports.ValidationExceptionReason = exports.ThrottlingException = exports.ServiceQuotaExceededException = exports.ResourceNotFoundException = exports.InternalServerException = exports.ConflictException = exports.ResourceType = exports.ExternalConnectionStatus = exports.PackageFormat = exports.HashAlgorithm = exports.AllowUpstream = exports.AllowPublish = exports.AccessDeniedException = void 0;
const CodeartifactServiceException_1 = require("./CodeartifactServiceException");
class AccessDeniedException extends CodeartifactServiceException_1.CodeartifactServiceException {
    constructor(opts) {
        super({
            name: "AccessDeniedException",
            $fault: "client",
            ...opts,
        });
        this.name = "AccessDeniedException";
        this.$fault = "client";
        Object.setPrototypeOf(this, AccessDeniedException.prototype);
    }
}
exports.AccessDeniedException = AccessDeniedException;
exports.AllowPublish = {
    ALLOW: "ALLOW",
    BLOCK: "BLOCK",
};
exports.AllowUpstream = {
    ALLOW: "ALLOW",
    BLOCK: "BLOCK",
};
exports.HashAlgorithm = {
    MD5: "MD5",
    SHA1: "SHA-1",
    SHA256: "SHA-256",
    SHA512: "SHA-512",
};
exports.PackageFormat = {
    GENERIC: "generic",
    MAVEN: "maven",
    NPM: "npm",
    NUGET: "nuget",
    PYPI: "pypi",
};
exports.ExternalConnectionStatus = {
    AVAILABLE: "Available",
};
exports.ResourceType = {
    ASSET: "asset",
    DOMAIN: "domain",
    PACKAGE: "package",
    PACKAGE_VERSION: "package-version",
    REPOSITORY: "repository",
};
class ConflictException extends CodeartifactServiceException_1.CodeartifactServiceException {
    constructor(opts) {
        super({
            name: "ConflictException",
            $fault: "client",
            ...opts,
        });
        this.name = "ConflictException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ConflictException.prototype);
        this.resourceId = opts.resourceId;
        this.resourceType = opts.resourceType;
    }
}
exports.ConflictException = ConflictException;
class InternalServerException extends CodeartifactServiceException_1.CodeartifactServiceException {
    constructor(opts) {
        super({
            name: "InternalServerException",
            $fault: "server",
            ...opts,
        });
        this.name = "InternalServerException";
        this.$fault = "server";
        Object.setPrototypeOf(this, InternalServerException.prototype);
    }
}
exports.InternalServerException = InternalServerException;
class ResourceNotFoundException extends CodeartifactServiceException_1.CodeartifactServiceException {
    constructor(opts) {
        super({
            name: "ResourceNotFoundException",
            $fault: "client",
            ...opts,
        });
        this.name = "ResourceNotFoundException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ResourceNotFoundException.prototype);
        this.resourceId = opts.resourceId;
        this.resourceType = opts.resourceType;
    }
}
exports.ResourceNotFoundException = ResourceNotFoundException;
class ServiceQuotaExceededException extends CodeartifactServiceException_1.CodeartifactServiceException {
    constructor(opts) {
        super({
            name: "ServiceQuotaExceededException",
            $fault: "client",
            ...opts,
        });
        this.name = "ServiceQuotaExceededException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ServiceQuotaExceededException.prototype);
        this.resourceId = opts.resourceId;
        this.resourceType = opts.resourceType;
    }
}
exports.ServiceQuotaExceededException = ServiceQuotaExceededException;
class ThrottlingException extends CodeartifactServiceException_1.CodeartifactServiceException {
    constructor(opts) {
        super({
            name: "ThrottlingException",
            $fault: "client",
            ...opts,
        });
        this.name = "ThrottlingException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ThrottlingException.prototype);
        this.retryAfterSeconds = opts.retryAfterSeconds;
    }
}
exports.ThrottlingException = ThrottlingException;
exports.ValidationExceptionReason = {
    CANNOT_PARSE: "CANNOT_PARSE",
    ENCRYPTION_KEY_ERROR: "ENCRYPTION_KEY_ERROR",
    FIELD_VALIDATION_FAILED: "FIELD_VALIDATION_FAILED",
    OTHER: "OTHER",
    UNKNOWN_OPERATION: "UNKNOWN_OPERATION",
};
class ValidationException extends CodeartifactServiceException_1.CodeartifactServiceException {
    constructor(opts) {
        super({
            name: "ValidationException",
            $fault: "client",
            ...opts,
        });
        this.name = "ValidationException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ValidationException.prototype);
        this.reason = opts.reason;
    }
}
exports.ValidationException = ValidationException;
exports.PackageVersionErrorCode = {
    ALREADY_EXISTS: "ALREADY_EXISTS",
    MISMATCHED_REVISION: "MISMATCHED_REVISION",
    MISMATCHED_STATUS: "MISMATCHED_STATUS",
    NOT_ALLOWED: "NOT_ALLOWED",
    NOT_FOUND: "NOT_FOUND",
    SKIPPED: "SKIPPED",
};
exports.PackageVersionStatus = {
    ARCHIVED: "Archived",
    DELETED: "Deleted",
    DISPOSED: "Disposed",
    PUBLISHED: "Published",
    UNFINISHED: "Unfinished",
    UNLISTED: "Unlisted",
};
exports.DomainStatus = {
    ACTIVE: "Active",
    DELETED: "Deleted",
};
exports.PackageVersionOriginType = {
    EXTERNAL: "EXTERNAL",
    INTERNAL: "INTERNAL",
    UNKNOWN: "UNKNOWN",
};
exports.PackageVersionSortType = {
    PUBLISHED_TIME: "PUBLISHED_TIME",
};
const GetPackageVersionAssetResultFilterSensitiveLog = (obj) => ({
    ...obj,
});
exports.GetPackageVersionAssetResultFilterSensitiveLog = GetPackageVersionAssetResultFilterSensitiveLog;
const PublishPackageVersionRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
exports.PublishPackageVersionRequestFilterSensitiveLog = PublishPackageVersionRequestFilterSensitiveLog;
